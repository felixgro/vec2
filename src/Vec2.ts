const clamp = (val: number, min: number, max: number): number => {
   return Math.min(Math.max(val, min), max);
};

export const distanceBetween = (vecA: Vec2, vecB: Vec2): number => {
   return vecA.distanceTo(vecB);
}

export const angleBetween = (vecA: Vec2, vecB: Vec2): number => {
   return vecA.angleTo(vecB);
}

export default class Vec2 {
   public x: number;
   public y: number;

   constructor(x = 0, y = 0) {
      this.x = x;
      this.y = y;
   }

   static up = (): Vec2 => new Vec2(0, -1);
   static down = (): Vec2 => new Vec2(0, 1);
   static left = (): Vec2 => new Vec2(-1, 0);
   static right = (): Vec2 => new Vec2(1, 0);

   static fromAngle(angle: number, length: number): Vec2 {
      return new Vec2();
   }

   static lerp(vecA: Vec2, vecB: Vec2, progress: number): Vec2 {
      if (progress > 1) progress = 1;
      if (progress < 0) progress = 0;

      const dirVec = vecB.clone().subtract(vecA);

      return vecA.clone().add(dirVec.multiply(progress));
   }

   get rawPosition(): [number, number] {
      return [this.x, this.y];
   }

   get length(): number {
      return Math.sqrt(this.x * this.x + this.y * this.y);
   }

   public add(vec: Vec2): Vec2 {
      this.x += vec.x;
      this.y += vec.y;

      return this;
   }

   public subtract(vec: Vec2): Vec2 {
      this.x -= vec.x;
      this.y -= vec.y;

      return this;
   }

   public multiply(n: number): Vec2 {
      this.x *= n;
      this.y *= n;

      return this;
   }

   public divide(n: number): Vec2 {
      if (n === 0) throw new Error('Cannot divide by 0');

      this.x /= n;
      this.y /= n;

      return this;
   }

   public normalize(): Vec2 {
      return this.divide(this.length);
   }

   public equals(vec: Vec2): boolean {
      return this.x === vec.x && this.y === vec.y;
   }

   public distanceTo(vec: Vec2): number {
      return vec.clone().subtract(this).length
   }

   public angleTo(vec: Vec2): number {
      let angle = Math.atan2(vec.x, vec.y) - Math.atan2(this.x, this.y);

      if (angle > Math.PI) angle -= Math.PI;
      if (angle < -Math.PI) angle += 2 * Math.PI;

      return angle;
   }

   public clone(): Vec2 {
      return new Vec2(this.x, this.y);
   }

   public drawOnCanvas(ctx: CanvasRenderingContext2D, options: {
      origin: Vec2;
      color: string;
      width: number;
   }): void {
      ctx.save();
      ctx.translate(...options.origin.rawPosition);

      ctx.strokeStyle = options.color;
      ctx.lineWidth = options.width;

      const arrowWidth = clamp(3 * options.width, 7, 50);
      const angle = Vec2.up().angleTo(this);

      ctx.beginPath();
      ctx.moveTo(0, 0);
      ctx.lineTo(...this.rawPosition);
      ctx.stroke();

      ctx.restore();
   }
}