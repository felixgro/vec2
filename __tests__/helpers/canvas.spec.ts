import Vec2, { draw, drawMany } from '../../src/index';

const getCtx = (): CanvasRenderingContext2D => {
   const canvas = document.createElement('canvas');
   return canvas.getContext('2d')!;
}

describe('draw', () => {
   test('with default options', () => {
      const vec = Vec2.up();
      const ctx = getCtx();

      draw(vec, ctx);

      // @ts-ignore
      expect(ctx.__getEvents()).toMatchSnapshot();
   });

   test('with custom options', () => {
      const vec = Vec2.up();
      const ctx = getCtx();

      draw(vec, ctx, {
         origin: new Vec2(100, 100),
         color: '#ff0000',
         width: 3,
         arrow: false
      });

      // @ts-ignore
      expect(ctx.__getEvents()).toMatchSnapshot();
   });
});

describe('drawMany', () => {
   test('with default options', () => {
      const vecA = Vec2.up();
      const vecB = Vec2.down();
      const ctx = getCtx();

      drawMany([vecA, vecB], ctx);

      // @ts-ignore
      expect(ctx.__getEvents()).toMatchSnapshot();
   });

   test('with custom options', () => {
      const vecA = Vec2.up();
      const vecB = Vec2.down();
      const ctx = getCtx();

      drawMany([vecA, vecB], ctx, {
         origin: new Vec2(100, 100),
         color: '#ff0000',
         width: 5,
         arrow: false
      });

      // @ts-ignore
      expect(ctx.__getEvents()).toMatchSnapshot();
   });
})