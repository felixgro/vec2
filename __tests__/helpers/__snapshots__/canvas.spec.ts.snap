// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw with custom options 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#ff0000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 100,
      "y": 100,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": -1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`draw with default options 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": -1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {
      "angle": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "rotate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 7,
      "y": 7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": -7,
      "y": 7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 7,
            "y": 7,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": -7,
            "y": 7,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`drawMany with custom options 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#ff0000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 100,
      "y": 100,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": -1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#ff0000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 100,
      "y": 100,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            100,
            100,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      100,
      100,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`drawMany with default options 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": -1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": -1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {
      "angle": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "rotate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 7,
      "y": 7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": -7,
      "y": 7,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 7,
            "y": 7,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": -7,
            "y": 7,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            -1,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      -1,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineCap",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 1,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      1,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {
      "angle": 3.141592653589793,
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "rotate",
  },
  Object {
    "props": Object {},
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 7,
      "y": 7,
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": -7,
      "y": 7,
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            -1,
            1.2246467991473532e-16,
            -1.2246467991473532e-16,
            -1,
            0,
            1,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            -1,
            1.2246467991473532e-16,
            -1.2246467991473532e-16,
            -1,
            0,
            1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 7,
            "y": 7,
          },
          "transform": Array [
            -1,
            1.2246467991473532e-16,
            -1.2246467991473532e-16,
            -1,
            0,
            1,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 0,
            "y": 0,
          },
          "transform": Array [
            -1,
            1.2246467991473532e-16,
            -1.2246467991473532e-16,
            -1,
            0,
            1,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": -7,
            "y": 7,
          },
          "transform": Array [
            -1,
            1.2246467991473532e-16,
            -1.2246467991473532e-16,
            -1,
            0,
            1,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      -1,
      1.2246467991473532e-16,
      -1.2246467991473532e-16,
      -1,
      0,
      1,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;
